src/main/java/com/samcomo/dbz/global/redis/aop/DistributedLockAop.java

```diff
@@ -0,0 +1,61 @@
+package com.samcomo.dbz.global.redis.aop;
+
+import com.samcomo.dbz.global.exception.ErrorCode;
+import com.samcomo.dbz.global.redis.LockType;
+import com.samcomo.dbz.global.redis.exception.RedissonException;
+import java.lang.reflect.Method;
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import org.aspectj.lang.ProceedingJoinPoint;
+import org.aspectj.lang.annotation.Around;
+import org.aspectj.lang.annotation.Aspect;
+import org.aspectj.lang.reflect.MethodSignature;
+import org.redisson.api.RLock;
+import org.redisson.api.RedissonClient;
+import org.springframework.stereotype.Component;
+
+@Aspect
+@Component
+@RequiredArgsConstructor
+@Slf4j
+public class DistributedLockAop {
+
+  private final RedissonClient redissonClient;
+  private final AopForTransaction aopForTransaction;
+
+  @Around("@annotation(com.samcomo.dbz.global.redis.aop.DistributedLock)")
+  public Object lock(ProceedingJoinPoint joinPoint) throws Throwable{
+
+    MethodSignature signature = (MethodSignature) joinPoint.getSignature();
+    Method method = signature.getMethod();
+    DistributedLock distributedLock = method.getAnnotation(DistributedLock.class);
+
+    String lockName = distributedLock.key();
+    LockType type = distributedLock.lockType();
+    RLock lock = redissonClient.getLock(lockName + ":" + type);
+
+    try{
+      if (!lock.tryLock(
+          distributedLock.waitTime(), distributedLock.leaseTime(), distributedLock.timeUnit())){
+        throw new RedissonException(ErrorCode.LOCK_FAIL);
+      }
+
+      log.info("Lock 획득!");
+
+      return aopForTransaction.proceed(joinPoint);
+
+    }catch (InterruptedException e){
+      throw new RedissonException(ErrorCode.LOCK_FAIL);
```

Comment:
```suggestion
    }catch (InterruptedException e){
      throw new RedissonException(ErrorCode.LOCK_INTERRUPTED);
```

tryLock() 에서 이미 점유되어 있는 락 때문에 락 획득을 실패한 경우의 에러처리와
내부 문제(락이 interrupt 되어서 에러가 난 상황)를 구분해서 예외처리하면 좋을 것 같아요.

Reply:
그리고 tryLock() 에서 락 얻는 거 실패하면 return false 처리하시는 분들도 많던데 에러로 던진 이유도 궁금합니다!

Reply:
> tryLock() 에서 이미 점유되어 있는 락 때문에 락 획득을 실패한 경우의 에러처리와
내부 문제(락이 interrupt 되어서 에러가 난 상황)를 구분해서 예외처리하면 좋을 것 같아요.
- 두 가지 상황이 있으니까 분리하는 방법이 더 좋겠네요!

> 그리고 tryLock() 에서 락 얻는 거 실패하면 return false 처리하시는 분들도 많던데 에러로 던진 이유도 궁금합니다!
return false로 해도 문제는 없지만 저는 에러난 상황을 확인하고 싶어서 에러로 던졌어요! waitTime이나 leaseTime 을 적합하지 않게 설정해서 동시성 처리가 제대로 해결되지 않았다면 에러 로그를 확인하고 바로 해결할 수 있도록 하고 싶었어요!